File: leetcode-go/Q001-Q100/Q074_Search_a_2D_Matrix/Q074_Search_a_2D_Matrix.go
package Q074_Search_a_2D_Matrix

func searchMatrix(matrix [][]int, target int) bool {
	if len(matrix) == 0 {
		return false
	}
	m, n := len(matrix), len(matrix[0])
	l, h := 0, m*n-1
	for l <= h {
		mid := (l + h) / 2
		i, j := mid/n, mid%n
		if matrix[i][j] == target {
			return true
		}
		if matrix[i][j] < target {
			l = mid + 1
		} else {
			h = mid - 1
		}
	}
	return false
}

File: leetcode-java/src/Q001_Q099/Q074_Search_a_2D_Matrix.java
package Q001_Q099;

public class Q074_Search_a_2D_Matrix {

    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0 || matrix[0] == null || matrix[0].length == 0) return false;
        int row = matrix.length, col = matrix[0].length;
        int i = 0, j = row*col-1;
        while (i <= j) {
            int m = (i+j)/2;
            if (matrix[m/col][m%col] > target) j = m-1;
            else if (matrix[m/col][m%col] < target) i = m+1;
            else return true;
        }
        return false;
    }

}
